File: crates/varve-effects/src/crt.rs
//! CRT / analog display emulation kernel — faithful port of
//! `packages/engine/src/liveEffects/crt.ts`.
//!
//! Reusable procedural primitives: barrel curvature warp, scanline shading,
//! phosphor mask layouts (RGB stripe / BGR stripe / aperture grille / shadow
//! mask), glow, vignette, sub-pixel convergence offsets, and brightness/
//! contrast gain. Every pattern is analytic and deterministic — no noise, no
//! randomness.

use crate::{clamp01, clamp_byte, js_round, Params};

fn lerp(a: f64, b: f64, t: f64) -> f64 {
    a + (b - a) * t
}

fn smoothstep(e0: f64, e1: f64, x: f64) -> f64 {
    let t = clamp01((x - e0) / (e1 - e0));
    t * t * (3.0 - 2.0 * t)
}

fn clamp_range(v: f64, lo: f64, hi: f64) -> f64 {
    v.max(lo).min(hi)
}

fn premultiply(data: &mut [u8]) {
    for i in (0..data.len()).step_by(4) {
        let a = data[i + 3];
        if a == 255 {
            continue;
        }
        if a == 0 {
            data[i] = 0;
            data[i + 1] = 0;
            data[i + 2] = 0;
            continue;
        }
        data[i] = clamp_byte((data[i] as f64 * a as f64) / 255.0);
        data[i + 1] = clamp_byte((data[i + 1] as f64 * a as f64) / 255.0);
        data[i + 2] = clamp_byte((data[i + 2] as f64 * a as f64) / 255.0);
    }
}

fn unpremultiply(data: &mut [u8]) {
    for i in (0..data.len()).step_by(4) {
        let a = data[i + 3];
        if a == 0 || a == 255 {
            continue;
        }
        let inv = 255.0 / a as f64;
        data[i] = clamp_byte(data[i] as f64 * inv);
        data[i + 1] = clamp_byte(data[i + 1] as f64 * inv);
        data[i + 2] = clamp_byte(data[i + 2] as f64 * inv);
    }
}

/// Sub-pixel bilinear sample of channel `c` — port of `sampleBilinear`.
fn sample_bilinear(src: &[u8], w: u32, h: u32, x: f64, y: f64, c: usize) -> f64 {
    let x0 = x.floor() as i64;
    let y0 = y.floor() as i64;
    let fx = x - x0 as f64;
    let fy = y - y0 as f64;
    let w_ = w as i64;
    let h_ = h as i64;
    let x1 = (x0 + 1).clamp(0, w_ - 1);
    let y1 = (y0 + 1).clamp(0, h_ - 1);
    let c0 = x0.clamp(0, w_ - 1);
    let r0 = y0.clamp(0, h_ - 1);
    let a = src[((r0 * w_ + c0) * 4 + c as i64) as usize] as f64;
    let b = src[((r0 * w_ + x1) * 4 + c as i64) as usize] as f64;
    let d = src[((y1 * w_ + c0) * 4 + c as i64) as usize] as f64;
    let e = src[((y1 * w_ + x1) * 4 + c as i64) as usize] as f64;
    a + (b - a) * fx + (d - a) * fy + (a - b - d + e) * fx * fy
}

/// Phosphor mask multiplier triple — port of `phosphorMaskAt`.
fn phosphor_mask_at(mask: &str, x: f64, y: f64, pitch: f64) -> (f64, f64, f64) {
    let px = x % pitch;
    let t = px / pitch;
    match mask {
        "rgb-stripe" => {
            if t < 0.34 {
                (1.0, 0.22, 0.22)
            } else if t < 0.67 {
                (0.22, 1.0, 0.22)
            } else {
                (0.22, 0.22, 1.0)
            }
        }
        "bgr-stripe" => {
            if t < 0.34 {
                (0.22, 0.22, 1.0)
            } else if t < 0.67 {
                (0.22, 1.0, 0.22)
            } else {
                (1.0, 0.22, 0.22)
            }
        }
        "aperture-grille" => {
            if t < 0.5 {
                (1.0, 0.35, 0.35)
            } else {
                (0.35, 0.35, 1.0)
            }
        }
        "shadow-mask" => {
            let py = y % pitch;
            let dot = (t - 0.5).hypot(py / pitch - 0.5) * 2.0;
            let dark = if dot > 0.85 { 0.2 } else { 1.0 };
            (dark, dark, dark)
        }
        _ => (1.0, 1.0, 1.0),
    }
}

/// 3×3-ish box blur with float32 accumulation — port of `boxBlur3`.
fn box_blur3(data: &mut [u8], w: u32, h: u32, radius: i64) {
    let mut tmp = vec![0.0f32; (w * h * 4) as usize];
    for c in 0..4u32 {
        for y in 0..h {
            for x in 0..w {
                let mut sum = 0.0f64;
                let mut n = 0.0f64;
                for dy in -radius..=radius {
                    let ny = y as i64 + dy;
                    if ny < 0 || ny >= h as i64 {
                        continue;
                    }
                    for dx in -radius..=radius {
                        let nx = x as i64 + dx;
                        if nx < 0 || nx >= w as i64 {
                            continue;
                        }
                        sum += data[((ny * w as i64 + nx) * 4 + c as i64) as usize] as f64;
                        n += 1.0;
                    }
                }
                tmp[((y * w + x) * 4 + c) as usize] = (sum / n) as f32;
            }
        }
    }
    for (i, t) in tmp.iter().enumerate() {
        data[i] = js_round(*t as f64) as u8;
    }
}

/// Apply CRT emulation in place — port of `applyCrt`.
pub fn apply(out: &mut [u8], w: u32, h: u32, p: &Params) {
    let curvature = clamp01(p.f("curvature", 0.0));
    let scan_period = p.f("scanlinePeriod", 3.0).max(1.5);
    let scan_strength = clamp01(p.f("scanlineStrength", 0.5));
    let scan_softness = clamp01(p.f("scanlineSoftness", 0.5));
    let phosphor = p.s("phosphorMask", "none");
    let pitch = p.f("phosphorPitch", 4.0).max(1.0);
    let ph_intensity = clamp01(p.f("phosphorIntensity", 0.6));
    let glow = clamp01(p.f("glow", 0.0));
    let vignette = clamp01(p.f("vignette", 0.0));
    let vignette_r = clamp01(p.f("vignetteRadius", 0.5));
    let cx = p.f("convergenceX", 0.0);
    let cy = p.f("convergenceY", 0.0);
    let brightness = clamp_range(p.f("brightness", 0.0), -1.0, 1.0);
    let contrast = clamp_range(p.f("contrast", 1.0), 0.0, 2.0);

    let mut src = out.to_vec();
    premultiply(&mut src);

    let half_w = w as f64 / 2.0;
    let half_h = h as f64 / 2.0;
    let warp_k = curvature * 0.28;
    let wf = w as f64;
    let hf = h as f64;
    let wu = w as usize;

    for y in 0..h {
        for x in 0..w {
            let o = (y as usize * wu + x as usize) * 4;
            let a = src[o + 3];

            // Curvature warp: inverse mapping (output px -> source offset).
            let mut sx = x as f64;
            let mut sy = y as f64;
            if warp_k > 0.0 {
                let nx = (x as f64 - half_w) / half_w;
                let ny = (y as f64 - half_h) / half_h;
                let r2 = nx * nx + ny * ny;
                let scale = 1.0 + warp_k * r2;
                sx = (half_w + (nx * half_w) / scale).max(0.0).min(wf - 1.0);
                sy = (half_h + (ny * half_h) / scale).max(0.0).min(hf - 1.0);
            }
            let x0 = sx.floor();
            let y0 = sy.floor();
            let fx = sx - x0;
            let fy = sy - y0;
            let x1 = (x0 + 1.0).min(wf - 1.0);
            let y1 = (y0 + 1.0).min(hf - 1.0);
            let x0i = x0 as usize;
            let y0i = y0 as usize;
            let x1i = x1 as usize;
            let y1i = y1 as usize;
            let o00 = (y0i * wu + x0i) * 4;
            let o10 = (y0i * wu + x1i) * 4;
            let o01 = (y1i * wu + x0i) * 4;
            let o11 = (y1i * wu + x1i) * 4;
            let mut r = lerp(
                lerp(src[o00] as f64, src[o10] as f64, fx),
                lerp(src[o01] as f64, src[o11] as f64, fx),
                fy,
            );
            let mut g = lerp(
                lerp(src[o00 + 1] as f64, src[o10 + 1] as f64, fx),
                lerp(src[o01 + 1] as f64, src[o11 + 1] as f64, fx),
                fy,
            );
            let mut b = lerp(
                lerp(src[o00 + 2] as f64, src[o10 + 2] as f64, fx),
                lerp(src[o01 + 2] as f64, src[o11 + 2] as f64, fx),
                fy,
            );

            // Convergence: red shifted +, blue shifted - (subpixel bilinear).
            if cx != 0.0 || cy != 0.0 {
                let rr = sample_bilinear(&src, w, h, x as f64 + cx, y as f64 + cy, 0);
                let rb = sample_bilinear(&src, w, h, x as f64 - cx, y as f64 - cy, 2);
                r = lerp(r, rr, 0.6);
                b = lerp(b, rb, 0.6);
            }

            // Scanlines.
            if scan_strength > 0.0 {
                let phase = ((y as f64 % scan_period) + scan_period) % scan_period;
                let pulse = 0.5 + 0.5 * (2.0 * std::f64::consts::PI * phase / scan_period).cos();
                let depth = scan_strength * pulse.powf(0.4 + scan_softness * 2.2);
                r *= 1.0 - depth;
                g *= 1.0 - depth;
                b *= 1.0 - depth;
            }

            // Phosphor mask.
            if phosphor != "none" && ph_intensity > 0.0 {
                let (mr, mg, mb) = phosphor_mask_at(phosphor, x as f64, y as f64, pitch);
                let m = ph_intensity;
                r = r * (1.0 - m) + r * mr * m;
                g = g * (1.0 - m) + g * mg * m;
                b = b * (1.0 - m) + b * mb * m;
            }

            // Vignette (also handles corner rounding via the same falloff).
            if vignette > 0.0 {
                let nx = (x as f64 - half_w) / (half_w * vignette_r * 2.0);
                let ny = (y as f64 - half_h) / (half_h * vignette_r * 2.0);
                let d = nx.hypot(ny).min(1.0);
                let vig = 1.0 - vignette * smoothstep(0.55, 1.0, d);
                r *= vig;
                g *= vig;
                b *= vig;
            }

            // Brightness/contrast.
            let gain = contrast;
            r = (r - 128.0) * gain + 128.0 + brightness * 128.0;
            g = (g - 128.0) * gain + 128.0 + brightness * 128.0;
            b = (b - 128.0) * gain + 128.0 + brightness * 128.0;

            out[o] = clamp_byte(r);
            out[o + 1] = clamp_byte(g);
            out[o + 2] = clamp_byte(b);
            out[o + 3] = a;
        }
    }

    // Glow: add a small blur of the bright content back (linear-light friendly).
    if glow > 0.0 {
        let mut glow_pass = out.to_vec();
        box_blur3(&mut glow_pass, w, h, 2);
        let m = glow * 0.5;
        for i in (0..out.len()).step_by(4) {
            out[i] = clamp_byte(out[i] as f64 + (glow_pass[i] as f64 - out[i] as f64) * m);
            out[i + 1] =
                clamp_byte(out[i + 1] as f64 + (glow_pass[i + 1] as f64 - out[i + 1] as f64) * m);
            out[i + 2] =
                clamp_byte(out[i + 2] as f64 + (glow_pass[i + 2] as f64 - out[i + 2] as f64) * m);
        }
    }

    unpremultiply(out);
}

File: packages/engine/src/liveEffects/crt.ts
/**
 * CRT / analog display emulation kernel.
 *
 * Reusable procedural primitives: barrel curvature warp, scanline shading,
 * phosphor mask layouts (RGB stripe / BGR stripe / aperture grille / shadow
 * mask), glow, vignette, sub-pixel convergence offsets, and brightness/
 * contrast gain. Every pattern is analytic and deterministic — no noise, no
 * randomness. Scanline/phosphor moiré is mitigated by sampling the mask at
 * device pixels while curvature is applied in normalized coordinates, so the
 * pattern only aliases at extreme zoom (documented; the saved effect is
 * unchanged by zoom).
 */

export type PhosphorMask = 'none' | 'rgb-stripe' | 'bgr-stripe' | 'aperture-grille' | 'shadow-mask';

export interface CrtParams {
  /** 0..1 barrel curvature. */
  curvature: number;
  /** 0..1 corner rounding (multiplies the vignette falloff). */
  cornerRadius: number;
  /** Scanline period in device pixels (>= 1.5). */
  scanlinePeriod: number;
  /** 0..1 scanline depth. */
  scanlineStrength: number;
  /** 0..1 scanline softness. */
  scanlineSoftness: number;
  phosphorMask: PhosphorMask;
  /** Phosphor pitch in device pixels (>= 1). */
  phosphorPitch: number;
  /** 0..1 phosphor mask depth. */
  phosphorIntensity: number;
  /** 0..1 phosphor glow (small blur added to the bright pass). */
  glow: number;
  /** 0..1 vignette depth. */
  vignette: number;
  /** 0..1 vignette radius (0.5 = screen centre at 50%). */
  vignetteRadius: number;
  /** Convergence offsets in device pixels (red/blue opposed). */
  convergenceX: number;
  convergenceY: number;
  /** -1..1 brightness shift. */
  brightness: number;
  /** 0..2 contrast gain. */
  contrast: number;
}

/** Apply CRT emulation in place. Returns the same ImageData. */
export function applyCrt(imageData: ImageData, params: CrtParams): ImageData {
  const { data, width: w, height: h } = imageData;
  if (w === 0 || h === 0) return imageData;

  const curvature = clamp01(params.curvature ?? 0);
  const scanPeriod = Math.max(1.5, params.scanlinePeriod ?? 3);
  const scanStrength = clamp01(params.scanlineStrength ?? 0.5);
  const scanSoftness = clamp01(params.scanlineSoftness ?? 0.5);
  const phosphor = params.phosphorMask ?? 'none';
  const pitch = Math.max(1, params.phosphorPitch ?? 4);
  const phIntensity = clamp01(params.phosphorIntensity ?? 0.6);
  const glow = clamp01(params.glow ?? 0);
  const vignette = clamp01(params.vignette ?? 0);
  const vignetteR = clamp01(params.vignetteRadius ?? 0.5);
  const cx = params.convergenceX ?? 0;
  const cy = params.convergenceY ?? 0;
  const brightness = clampRange(params.brightness ?? 0, -1, 1);
  const contrast = clampRange(params.contrast ?? 1, 0, 2);

  const src = new Uint8ClampedArray(data);
  premultiply(src);

  const halfW = w / 2;
  const halfH = h / 2;

  // Curvature warp: inverse mapping (output px → source offset).
  // Barrel: out_r = in_r * (1 + k * in_r²) → sample at in = out/(1+k·out²).
  const warpK = curvature * 0.28;

  // Phosphor mask pattern: evaluated analytically per subpixel phase.

  for (let y = 0; y < h; y += 1) {
    for (let x = 0; x < w; x += 1) {
      const o = (y * w + x) * 4;
      const a = src[o + 3]!;

      // Curvature: sample source at warped position (clamp border).
      let sx = x;
      let sy = y;
      if (warpK > 0) {
        const nx = (x - halfW) / halfW;
        const ny = (y - halfH) / halfH;
        const r2 = nx * nx + ny * ny;
        const scale = 1 + warpK * r2;
        sx = Math.min(w - 1, Math.max(0, halfW + (nx * halfW) / scale));
        sy = Math.min(h - 1, Math.max(0, halfH + (ny * halfH) / scale));
      }
      const x0 = Math.floor(sx);
      const y0 = Math.floor(sy);
      const fx = sx - x0;
      const fy = sy - y0;
      const x1 = Math.min(w - 1, x0 + 1);
      const y1 = Math.min(h - 1, y0 + 1);
      let r = lerp(
        lerp(src[(y0 * w + x0) * 4]!, src[(y0 * w + x1) * 4]!, fx),
        lerp(src[(y1 * w + x0) * 4]!, src[(y1 * w + x1) * 4]!, fx),
        fy,
      );
      let g = lerp(
        lerp(src[(y0 * w + x0) * 4 + 1]!, src[(y0 * w + x1) * 4 + 1]!, fx),
        lerp(src[(y1 * w + x0) * 4 + 1]!, src[(y1 * w + x1) * 4 + 1]!, fx),
        fy,
      );
      let b = lerp(
        lerp(src[(y0 * w + x0) * 4 + 2]!, src[(y0 * w + x1) * 4 + 2]!, fx),
        lerp(src[(y1 * w + x0) * 4 + 2]!, src[(y1 * w + x1) * 4 + 2]!, fx),
        fy,
      );

      // Convergence: red shifted +, blue shifted − (subpixel bilinear).
      if (cx !== 0 || cy !== 0) {
        const rr = sampleBilinear(src, w, h, x + cx, y + cy, 0);
        const rb = sampleBilinear(src, w, h, x - cx, y - cy, 2);
        r = lerp(r, rr, 0.6);
        b = lerp(b, rb, 0.6);
      }

      // Scanlines.
      if (scanStrength > 0) {
        const phase = ((y % scanPeriod) + scanPeriod) % scanPeriod;
        const pulse = 0.5 + 0.5 * Math.cos((2 * Math.PI * phase) / scanPeriod);
        const depth = scanStrength * pulse ** (0.4 + scanSoftness * 2.2);
        r *= 1 - depth;
        g *= 1 - depth;
        b *= 1 - depth;
      }

      // Phosphor mask.
      if (phosphor !== 'none' && phIntensity > 0) {
        const [mr, mg, mb] = phosphorMaskAt(phosphor, x, y, pitch);
        const m = phIntensity;
        r = r * (1 - m) + r * mr * m;
        g = g * (1 - m) + g * mg * m;
        b = b * (1 - m) + b * mb * m;
      }

      // Vignette (also handles corner rounding via the same falloff).
      if (vignette > 0) {
        const nx = (x - halfW) / (halfW * vignetteR * 2);
        const ny = (y - halfH) / (halfH * vignetteR * 2);
        const d = Math.min(1, Math.hypot(nx, ny));
        const vig = 1 - vignette * smoothstep(0.55, 1, d);
        r *= vig;
        g *= vig;
        b *= vig;
      }

      // Brightness/contrast.
      const gain = contrast;
      r = (r - 128) * gain + 128 + brightness * 128;
      g = (g - 128) * gain + 128 + brightness * 128;
      b = (b - 128) * gain + 128 + brightness * 128;

      data[o] = clampByte(r);
      data[o + 1] = clampByte(g);
      data[o + 2] = clampByte(b);
      data[o + 3] = a;
    }
  }

  // Glow: add a small blur of the bright content back (linear-light friendly).
  if (glow > 0) {
    const glowPass = new Uint8ClampedArray(data);
    boxBlur3(glowPass, w, h, 2);
    const m = glow * 0.5;
    for (let i = 0; i < data.length; i += 4) {
      data[i] = clampByte(data[i]! + (glowPass[i]! - data[i]!) * m);
      data[i + 1] = clampByte(data[i + 1]! + (glowPass[i + 1]! - data[i + 1]!) * m);
      data[i + 2] = clampByte(data[i + 2]! + (glowPass[i + 2]! - data[i + 2]!) * m);
    }
  }

  unpremultiply(data);
  return imageData;
}

function phosphorMaskAt(
  mask: PhosphorMask,
  x: number,
  y: number,
  pitch: number,
): [number, number, number] {
  const px = x % pitch;
  const t = px / pitch;
  switch (mask) {
    case 'rgb-stripe':
      return t < 0.34 ? [1, 0.22, 0.22] : t < 0.67 ? [0.22, 1, 0.22] : [0.22, 0.22, 1];
    case 'bgr-stripe':
      return t < 0.34 ? [0.22, 0.22, 1] : t < 0.67 ? [0.22, 1, 0.22] : [1, 0.22, 0.22];
    case 'aperture-grille': {
      // Vertical slots: narrow dark bands between columns of phosphors.
      return t < 0.5 ? [1, 0.35, 0.35] : [0.35, 0.35, 1];
    }
    case 'shadow-mask': {
      const py = y % pitch;
      const dot = Math.hypot(t - 0.5, py / pitch - 0.5) * 2;
      const dark = dot > 0.85 ? 0.2 : 1;
      return [dark, dark, dark];
    }
    default:
      return [1, 1, 1];
  }
}

function sampleBilinear(
  src: Uint8ClampedArray,
  w: number,
  h: number,
  x: number,
  y: number,
  c: number,
): number {
  const x0 = Math.floor(x);
  const y0 = Math.floor(y);
  const fx = x - x0;
  const fy = y - y0;
  const x1 = Math.min(w - 1, Math.max(0, x0 + 1));
  const y1 = Math.min(h - 1, Math.max(0, y0 + 1));
  const c0 = Math.max(0, Math.min(w - 1, x0));
  const r0 = Math.max(0, Math.min(h - 1, y0));
  const a = src[(r0 * w + c0) * 4 + c]!;
  const b = src[(r0 * w + x1) * 4 + c]!;
  const d = src[(y1 * w + c0) * 4 + c]!;
  const e = src[(y1 * w + x1) * 4 + c]!;
  return a + (b - a) * fx + (d - a) * fy + (a - b - d + e) * fx * fy;
}

function boxBlur3(data: Uint8ClampedArray, w: number, h: number, radius: number): void {
  const tmp = new Float32Array(w * h * 4);
  for (let c = 0; c < 4; c += 1) {
    for (let y = 0; y < h; y += 1) {
      for (let x = 0; x < w; x += 1) {
        let sum = 0;
        let n = 0;
        for (let dy = -radius; dy <= radius; dy += 1) {
          for (let dx = -radius; dx <= radius; dx += 1) {
            const nx = x + dx;
            const ny = y + dy;
            if (nx < 0 || nx >= w || ny < 0 || ny >= h) continue;
            sum += data[(ny * w + nx) * 4 + c]!;
            n += 1;
          }
        }
        tmp[(y * w + x) * 4 + c] = sum / n;
      }
    }
  }
  for (let i = 0; i < data.length; i += 1) data[i] = Math.round(tmp[i]!);
}

function premultiply(data: Uint8ClampedArray): void {
  for (let i = 0; i < data.length; i += 4) {
    const a = data[i + 3]!;
    if (a === 255) continue;
    if (a === 0) {
      data[i] = 0;
      data[i + 1] = 0;
      data[i + 2] = 0;
      continue;
    }
    data[i] = Math.round((data[i]! * a) / 255);
    data[i + 1] = Math.round((data[i + 1]! * a) / 255);
    data[i + 2] = Math.round((data[i + 2]! * a) / 255);
  }
}

function unpremultiply(data: Uint8ClampedArray): void {
  for (let i = 0; i < data.length; i += 4) {
    const a = data[i + 3]!;
    if (a === 0 || a === 255) continue;
    const inv = 255 / a;
    data[i] = clampByte(data[i]! * inv);
    data[i + 1] = clampByte(data[i + 1]! * inv);
    data[i + 2] = clampByte(data[i + 2]! * inv);
  }
}

function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

function smoothstep(e0: number, e1: number, x: number): number {
  const t = clamp01((x - e0) / (e1 - e0));
  return t * t * (3 - 2 * t);
}

function clampByte(v: number): number {
  return Math.max(0, Math.min(255, Math.round(v)));
}

function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}

function clampRange(v: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, v));
}
